### ISSUE TITLE
E3-4: Implement structured mutation strategy (no API key)

### ISSUE THREAD

**Issue by jkbrooks** [jkbrooks/autoresearch_network#12]
## E3-4: Implement structured mutation strategy (no API key required)

**Parent Epic:** #8 — Miner Implementation  
**Dependencies:** None (standalone module, uses only stdlib + `ast`)  
**Implementation order:** 4 of 9 (MVP)

---

### What This Is

A local, deterministic mutation engine that proposes modifications to `train.py` without requiring an external LLM API. This is the default strategy for miners who do not configure an API key, and the bootstrap path for testing the full miner stack before E3-6 is complete.

### Where It Goes

`autoresearch/mutations.py`

### Why This Exists

Not every miner will have an LLM API key. Not every test run should require API calls. The structured strategy provides a working, zero-cost mutation path that exercises the same pipeline as the LLM strategy. It also serves as the fallback when an LLM API call fails.

### Technical Context

From reading `train.py` (https://github.com/karpathy/autoresearch/blob/master/train.py), the highest-signal hyperparameters are:

| Knob | Location | Impact |
|------|----------|--------|
| `n_layer` / `DEPTH` | `GPTConfig` dataclass (~line 35) | Primary model complexity. More layers → potentially lower `val_bpb` but higher VRAM |
| `n_head`, `n_embd` | `GPTConfig` dataclass | Width. Wider → more params per layer |
| Learning rate | Optimizer block (~line 400 area) | Directly affects convergence speed |
| `TOTAL_BATCH_SIZE` | Module-level constant | Affects gradient estimation quality |
| `window_pattern` | `GPTConfig` dataclass | `"SSSL"` uses banded attention; `"L"` is simpler and may be faster on smaller GPUs |

### Interface

```python
from abc import ABC, abstractmethod

class MutationStrategy(ABC):
    @abstractmethod
    def propose(self, baseline_train_py: str) -> str:
        """Return a modified train.py source, or the original if no mutation available."""
        ...

class StructuredMutationStrategy(MutationStrategy):
    def __init__(self, seed: int = 42):
        self.tried: set[str] = set()
        self.rng = random.Random(seed)

    def propose(self, baseline_train_py: str) -> str:
        """Apply one untried mutation. Returns original unchanged when all exhausted."""
        ...

    @property
    def mutations_remaining(self) -> int:
        ...
```

### Built-in Mutations

Each mutation is a named transformation. They are tried in a shuffled order (seeded for determinism):

| Mutation name | What it does | Regex/replacement approach |
|---|---|---|
| `depth_increase` | `n_layer: int = N` → `n_layer: int = N+2` | Regex on `n_layer:\s*int\s*=\s*(\d+)` |
| `depth_decrease` | `n_layer: int = N` → `n_layer: int = max(4, N-2)` | Same regex, decrease |
| `lr_scale_up` | Multiply the first learning rate float by 1.5 | Find first `lr` or `learning_rate` assignment |
| `lr_scale_down` | Multiply the first learning rate float by 0.7 | Same target |
| `window_to_simple` | `window_pattern: str = "SSSL"` → `window_pattern: str = "L"` | Direct string replace |
| `window_to_sssl` | `window_pattern: str = "L"` → `window_pattern: str = "SSSL"` | Direct string replace |
| `batch_double` | `TOTAL_BATCH_SIZE = N` → `TOTAL_BATCH_SIZE = N * 2` | Regex on assignment |
| `batch_halve` | `TOTAL_BATCH_SIZE = N` → `TOTAL_BATCH_SIZE = N // 2` | Same |
| `embd_increase` | `n_embd: int = N` → `n_embd: int = N + 128` | Regex on `n_embd` |
| `embd_decrease` | `n_embd: int = N` → `n_embd: int = max(256, N - 128)` | Same |

### Implementation Notes

- Each mutation function takes the source string and returns a modified string.
- After applying a mutation, validate the output with `ast.parse()`. If parsing fails, skip that mutation and try the next.
- `tried` set tracks mutation names, not source hashes. Once `depth_increase` is tried, it is not re-tried even if the source changes.
- When `propose()` is called and all mutations in the shuffled list have been tried, return the original source unchanged. The caller (E3-5) checks for identity to detect exhaustion.
- The `MutationStrategy` base class is used by E3-6 (`LLMMutationStrategy`) as well, so the interface must be stable.

### Acceptance Criteria

- [ ] `StructuredMutationStrategy.propose(train_py)` returns a modified string that differs from input (on first call)
- [ ] After cycling through all mutations, `propose()` returns original unchanged
- [ ] Each mutation is applied only once per strategy instance
- [ ] Every mutation output passes `ast.parse()` (valid Python)
- [ ] `mutations_remaining` property reflects correct count
- [ ] Deterministic: same seed + same input → same sequence of mutations

### Test Expectations

- `test_depth_increase_mutation` — assert `n_layer` value is incremented in output
- `test_depth_decrease_mutation` — assert `n_layer` value is decremented (and not below 4)
- `test_lr_scale_mutations` — assert LR float is scaled up / down
- `test_window_pattern_mutation` — assert pattern string changes
- `test_batch_size_mutation` — assert TOTAL_BATCH_SIZE changes
- `test_embd_mutation` — assert n_embd changes
- `test_all_mutations_produce_valid_python` — loop through all mutations, `ast.parse()` each output
- `test_mutations_cycle_without_repeating` — call `propose()` N times, assert no mutation name appears twice
- `test_mutations_exhausted_returns_original` — after all tried, `propose()` returns identical string
- `test_deterministic_with_seed` — two instances with same seed produce identical mutation sequence
- `test_mutations_remaining_decrements` — assert property decreases by 1 after each `propose()`
